src/coreclr/src/System.Private.CoreLib/src/System/StubHelpers.Unix.cs

```diff
@@ -0,0 +1,45 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+
+internal static partial class Interop
+{
+    internal static partial class OleAut32
```

Comment:
It should be factored like the other BSTR function and just be in Marshal type. I do not think it makes sense to make this look like OleAut32 emulator for Unix

Reply:
Agree, I've moved to Marshal class and call it from stub helper in platform independent manner.
Investigating why `AnsiBStrTest` is currently throwing OOM like this:

```gdb
Process 28621 launched: '/Users/am11/projects/runtime_pr/artifacts/tests/coreclr/OSX.x64.Checked/Tests/Core_Root/corerun' (x86_64)
Out of memory.
Process 28621 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
    frame #0: 0x00007fff77a8f2c2 libsystem_kernel.dylib`__pthread_kill + 10
libsystem_kernel.dylib`__pthread_kill:
->  0x7fff77a8f2c2 <+10>: jae    0x7fff77a8f2cc            ; <+20>
    0x7fff77a8f2c4 <+12>: movq   %rax, %rdi
    0x7fff77a8f2c7 <+15>: jmp    0x7fff77a89453            ; cerror_nocancel
    0x7fff77a8f2cc <+20>: retq   
Target 0: (corerun) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
  * frame #0: 0x00007fff77a8f2c2 libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff77b4abf1 libsystem_pthread.dylib`pthread_kill + 284
    frame #2: 0x00007fff779f96a6 libsystem_c.dylib`abort + 127
    frame #3: 0x00000001018819de libcoreclr.dylib`::PROCAbort() at process.cpp:3474:5 [opt]
    frame #4: 0x000000010188175e libcoreclr.dylib`PROCEndProcess(hProcess=<unavailable>, uExitCode=1, bTerminateUnconditionally=YES) at process.cpp:1472:13 [opt]
    frame #5: 0x0000000101881803 libcoreclr.dylib`::TerminateProcess(hProcess=0x00000000ffffff01, uExitCode=1) at process.cpp:1371:11 [opt]
    frame #6: 0x0000000101c0832c libcoreclr.dylib`UnwindManagedExceptionPass1(ex=0x00007ffeefbfbbc0, frameContext=0x00007ffeefbfb660) at exceptionhandling.cpp:4614:17 [opt]
    frame #7: 0x0000000101c0852a libcoreclr.dylib`DispatchManagedException(ex=0x00007ffeefbfbbc0, isHardwareException=<unavailable>) at exceptionhandling.cpp:4686:17 [opt]
    frame #8: 0x00000001019f0747 libcoreclr.dylib`::PreStubWorker(pTransitionBlock=<unavailable>, pMD=0x00000001206f88a0) at prestub.cpp:1990:9 [opt]
    frame #9: 0x0000000101cb2643 libcoreclr.dylib`ThePreStub at theprestubamd64.S:16
    frame #10: 0x0000000120128d96
    frame #11: 0x0000000120128d0c
    frame #12: 0x000000012075923b
    frame #13: 0x0000000101cb1889 libcoreclr.dylib`CallDescrWorkerInternal at calldescrworkeramd64.S:97
    frame #14: 0x0000000101a78e4b libcoreclr.dylib`CallDescrWorkerWithHandler(pCallDescrData=0x0000000000000001, fCriticalCall=240) at callhelpers.cpp:68:5 [opt]
    frame #15: 0x0000000101a7990d libcoreclr.dylib`MethodDescCallSite::CallTargetWorker(this=0x00007ffeefbfeaa0, pArguments=0x00007ffeefbfec78, pReturnValue=0x00007ffeefbfec60, cbReturnValue=8) at callhelpers.cpp:544:9 [opt]
    frame #16: 0x00000001018b2936 libcoreclr.dylib`RunMain(MethodDesc*, short, int*, REF<PtrArray>*) [inlined] MethodDescCallSite::Call_RetArgSlot(this=0x0000000120965df0, pArguments=0x0000000000000000) at callhelpers.h:458:9 [opt]
    frame #17: 0x00000001018b2901 libcoreclr.dylib`RunMain(MethodDesc*, short, int*, REF<PtrArray>*) at assembly.cpp:1468 [opt]
    frame #18: 0x00000001018b26f3 libcoreclr.dylib`RunMain(MethodDesc*, short, int*, REF<PtrArray>*) [inlined] RunMain(pParam=<unavailable>)::$_0::operator()(Param*) const::'lambda'(Param*)::operator()(Param*) const at assembly.cpp:1536 [opt]
    frame #19: 0x00000001018b26f3 libcoreclr.dylib`RunMain(MethodDesc*, short, int*, REF<PtrArray>*) at assembly.cpp:1538 [opt]
    frame #20: 0x00000001018b2684 libcoreclr.dylib`RunMain(pFD=<unavailable>, numSkipArgs=1, piRetVal=<unavailable>, stringArgs=<unavailable>) at assembly.cpp:1538 [opt]
    frame #21: 0x00000001018b2d2f libcoreclr.dylib`Assembly::ExecuteMainMethod(this=<unavailable>, stringArgs=0x00007ffeefbff028, waitForOtherThreads=YES) at assembly.cpp:1648:18 [opt]
    frame #22: 0x0000000101910965 libcoreclr.dylib`CorHost2::ExecuteAssembly(this=<unavailable>, dwAppDomainId=<unavailable>, pwzAssemblyPath=<unavailable>, argc=<unavailable>, argv=0x0000000000000000, pReturnValue=<unavailable>) at corhost.cpp:384:39 [opt]
    frame #23: 0x0000000101891fb8 libcoreclr.dylib`::coreclr_execute_assembly(hostHandle=0x0000000100609050, domainId=<unavailable>, argc=<unavailable>, argv=<unavailable>, managedAssemblyPath=<unavailable>, exitCode=0x00007ffeefbff1ec) at unixinterface.cpp:431:24 [opt]
    frame #24: 0x0000000100002e94 corerun`ExecuteManagedAssembly(currentExeAbsolutePath="/Users/am11/projects/runtime_pr/artifacts/tests/coreclr/OSX.x64.Checked/Tests/Core_Root/corerun", clrFilesAbsolutePath=<unavailable>, managedAssemblyAbsolutePath="/Users/am11/projects/runtime_pr/artifacts/tests/coreclr/OSX.x64.Checked/Interop/StringMarshalling/AnsiBSTR/AnsiBStrTest/AnsiBStrTest.dll", managedAssemblyArgc=0, managedAssemblyArgv=0x0000000000000000) at coreruncommon.cpp:506:22 [opt]
    frame #25: 0x0000000100001e4c corerun`main(argc=<unavailable>, argv=<unavailable>) at corerun.cpp:148:20 [opt]
    frame #26: 0x00007fff779543d5 libdyld.dylib`start + 1
```

probably because it is now allocated on a different heap? CI was previously showing:

```
        /Users/runner/work/1/s/artifacts/tests/coreclr/OSX.x64.Checked/TestWrappers/Interop.StringMarshalling/Interop.StringMarshalling.XUnitWrapper.cs(140,0): at Interop_StringMarshalling._AnsiBSTR_AnsiBStrTest_AnsiBStrTest_._AnsiBSTR_AnsiBStrTest_AnsiBStrTest_sh()
      Output:
        
        Assert failure(PID 37883 [0x000093fb], Thread: 208440 [0x32e38]): !"Heap contamination detected! HeapFree was called on a heap other than the one that memory was allocated from.\n" "Possible cause: you used new (executable) to allocate the memory, but didn't use DeleteExecutable() to free it."
            File: /Users/runner/work/1/s/src/coreclr/src/utilcode/clrhost_nodependencies.cpp Line: 296
            Image: /tmp/helix/working/A73B09A0/p/corerun
```

but locally i did not get this assertion failure.

Reply:
> probably because it is now allocated on a different heap?

You code has buffer overruns.

Reply:
Thanks, found the rootcause of buffer overruns. Unfortunately, we cannot reuse AllocBSTR method as it allocates one additional bytes in case byteLength was an odd number. Since we do not have interop malloc for Windows, I had to move it out of Marshal.cs to target specific files.